{
 "cells": [
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-15T22:45:31.122379162Z",
     "start_time": "2025-12-15T22:45:12.509278899Z"
    }
   },
   "source": [
    "from CoDIAC import featureTools, InterPro, UniProt, PDB, IntegrateStructure_Reference, PTM, translateFeatures, \\\n",
    "    jalviewFunctions\n",
    "import CoDIAC\n",
    "import pandas as pd\n",
    "\n",
    "Interpro_ID = 'IPR000719'  # Prot_kinase_dom\n",
    "\n",
    "#We will be creating a lot of files, this is how we would like them to be named\n",
    "data_root = 'Data/'\n",
    "name_root = 'Prot_kinase_dom_' + Interpro_ID\n",
    "\n",
    "# The files we will make in this process (so that different pieces of code can be run below as needed)\n",
    "uniprot_reference_file = data_root + 'Uniprot_Reference/' + name_root + '_uniprot_reference.csv'  # The uniprot reference\n",
    "fasta_long_header_file = data_root + 'Uniprot_Reference/' + name_root + '_long_header.fasta'\n",
    "fasta_file = data_root + 'Uniprot_Reference/' + name_root + '.fasta'\n",
    "#note: in addition to these 3 files, this also makes a mapping file for movng between fasta_long_header_file and fasta_file\n",
    "\n",
    "#PDB Files we'll make in this process\n",
    "PDB_file = data_root + 'PDB_Reference/' + name_root + '_PDB.csv'\n",
    "PDB_file_annotated = data_root + 'PDB_Reference/' + name_root + '_PDB_annotated.csv'\n",
    "PDB_file_filtered = data_root + 'PDB_Reference/' + name_root + '_PDB_reference.csv'  #The final PDB structure file, containing only filtered proteins\n",
    "\n",
    "# PTMs feature directory location\n",
    "feature_dir = 'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features/'\n",
    "\n",
    "# Offsets for N and C termini\n",
    "N_OFFSET = 0\n",
    "C_OFFSET = -1\n"
   ],
   "outputs": [],
   "execution_count": 2
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 1: Get all Uniprot IDs that match a family of interest\n"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-15T20:32:17.746066386Z",
     "start_time": "2025-12-15T20:31:08.021185509Z"
    }
   },
   "source": [
    "uniprot_IDs, species_dict = CoDIAC.InterPro.fetch_uniprotids(Interpro_ID, REVIEWED=True, species='Homo sapiens')"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Found next page and downloading https://www.ebi.ac.uk/interpro/api/protein/reviewed/entry/interpro/IPR000719/?cursor=source%7Cs%7Cq00534&page_size=200&search=Homo+sapiens\n",
      "Found next page and downloading https://www.ebi.ac.uk/interpro/api/protein/reviewed/entry/interpro/IPR000719/?cursor=source%7Cs%7Cq99986&page_size=200&search=Homo+sapiens\n",
      "Fetched 481 Uniprot IDs linked to IPR000719, where count expected to be 481\n"
     ]
    }
   ],
   "execution_count": 2
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 2: Make a human reference file of the family of interest"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-15T22:00:01.592832451Z",
     "start_time": "2025-12-15T20:32:21.801092178Z"
    }
   },
   "source": [
    "\n",
    "uniprot_df = CoDIAC.UniProt.makeRefFile(uniprot_IDs, uniprot_reference_file)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Domain Reference File successfully created!\n",
      "Adding Interpro Domains\n",
      "Fetching domains..\n",
      "Appending domains to file..\n",
      "Interpro metadata succesfully incorporated\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Manually checked Uniprot reference file for errors/issues. Found Uniprot and Interpro generally agreeing, with Interpro adding additional domains of interest. However, the atypical SH2 domains (JAK family) differ quite a bit in their boundaries. After testing alignment effects, we selected the SMART domain boundaries, changing the InterPro boundaries by hand (since this is the column used) (SMART landed somehwere between the Uniprot defined boundaries and the longer boundaries of the InterPro database in length)\n",
    "\n",
    "Changes:\n",
    "1. JAK1 SH2:IPR000980:437:546 -> SH2:IPR000980:437:544; JAK2 SH2:IPR000980:397:501 -> 397:487; JAK3 SH2:IPR000980:373:477 -> 373:463; TYK2 SH2:IPR000980:450:553 -> 452:539 \n",
    "2. Manually removed the alpha-helix region termed \"PI3K_P85_iSH2:IPR032498\" in the Interpro file for P27986,PIK3R1; Q92569,PIK3R3; O00459,PIK3R2\n",
    "3. Manually removed C lobe SH2 domain in Supt6h, which overlaps with the parent SH2 domain:\n",
    "Spt6_SH2_C:IPR035018:1424:1515 (removed)\n"
   ]
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 3: Get information about all PDB IDs that exist for the reference proteins of interest"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "jupyter": {
     "is_executing": true
    },
    "ExecuteTime": {
     "end_time": "2025-12-16T02:00:07.207375484Z",
     "start_time": "2025-12-15T22:46:06.141653584Z"
    }
   },
   "source": [
    "\n",
    "CoDIAC.PDB.generateStructureRefFile_fromUniprotFile(uniprot_reference_file, PDB_file)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Structure Reference File successfully created!\n",
      "Could not fetch the following PDBs, these encountered errors, despite retrying 5 times:\n",
      "['3LM0', '3UVR']\n"
     ]
    }
   ],
   "execution_count": 3
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 4: Annotate the structure file with reference, for domain annotation"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-16T02:44:31.888452487Z",
     "start_time": "2025-12-16T02:13:00.606091685Z"
    }
   },
   "source": [
    "struct_df_out = CoDIAC.IntegrateStructure_Reference.add_reference_info_to_struct_file(PDB_file, uniprot_reference_file,\n",
    "                                                                                      PDB_file_annotated, INTERPRO=True,\n",
    "                                                                                      verbose=False)"
   ],
   "outputs": [],
   "execution_count": 4
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### STEP 5: Reduce the structure file to just those that contain the domain of interest"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-16T02:50:06.211012773Z",
     "start_time": "2025-12-16T02:50:03.813114170Z"
    }
   },
   "source": [
    "# Now with an appended PDB File, create an output that contains only the lines that have Prot_kinase_dom domains\n",
    "CoDIAC.IntegrateStructure_Reference.filter_structure_file(PDB_file_annotated, Interpro_ID, PDB_file_filtered)"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Made Data/PDB_Reference/Prot_kinase_dom_IPR000719_PDB_reference.csv file: 6360 structures retained of 7586 total\n"
     ]
    }
   ],
   "execution_count": 5
  },
  {
   "attachments": {},
   "cell_type": "markdown",
   "metadata": {},
   "source": "### STEP 6: Create the FASTA Reference file for Prot_kinase_dom domains"
  },
  {
   "cell_type": "code",
   "metadata": {
    "ExecuteTime": {
     "end_time": "2025-12-16T02:50:41.797080092Z",
     "start_time": "2025-12-16T02:50:41.668123917Z"
    }
   },
   "source": [
    "# Given the Prot_kinase_dom domain file, create the fasta reference file (using INTERPRO as default)\n",
    "\n",
    "CoDIAC.UniProt.print_domain_fasta_file(uniprot_reference_file, Interpro_ID, fasta_long_header_file, N_OFFSET, C_OFFSET,\n",
    "                                       APPEND=False)\n",
    "\n",
    "# Shortening the fasta headers, still unique for each domain/protein pair\n",
    "# dropping the redundant information about the domains printed \n",
    "\n",
    "key_array_order = ['uniprot', 'gene', 'domain_num', 'start', 'end']\n",
    "#translation creates a mapping file \n",
    "output_fasta, mapping_file = CoDIAC.UniProt.translate_fasta_to_new_headers(fasta_long_header_file, fasta_file,\n",
    "                                                                           key_array_order)\n"
   ],
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "n offset is 0 and c offset is -1\n",
      "Created files: Data/Uniprot_Reference/Prot_kinase_dom_IPR000719.fasta and Data/Uniprot_Reference/Prot_kinase_dom_IPR000719_mapping.csv\n"
     ]
    }
   ],
   "execution_count": 6
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "Perform Promals3D alignment at http://prodata.swmed.edu/promals3d/promals3d.php\n",
    "Using the fasta_file with shorter headers \n",
    "Once complete, replace the _ with | in the promals3d results\n",
    "Committed that to the alignment file location (see next cell where alignment file can be instatiated for later steps)\n"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# alignment file location, once created on Promals3D\n",
    "alignment_file = 'Data/Uniprot_Reference/alignment/Prot_kinase_dom_IPR000719_promals3d.fasta'\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "## STEP 7 Create the PTM feature files"
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### STEP 7a create the ProteomeScout based Features files"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "feature_dir_prot = feature_dir + '/ProteomeScout/'\n",
    "ptm_feature_file_list, ptm_count_dict, ptm_feature_dict, mapping_dict = CoDIAC.PTM.write_PTM_features(Interpro_ID,\n",
    "                                                                                                      uniprot_reference_file,\n",
    "                                                                                                      feature_dir_prot,\n",
    "                                                                                                      mapping_file,\n",
    "                                                                                                      N_OFFSET,\n",
    "                                                                                                      C_OFFSET,\n",
    "                                                                                                      gap_threshold=0.7,\n",
    "                                                                                                      num_PTM_threshold=5)\n",
    "print(\"Wrote these feature files:\")\n",
    "print(ptm_feature_file_list)\n",
    "print(\"These belong to the following fasta file:\")\n",
    "print(output_fasta)  #comes from block above - the short header format of the fasta header\n",
    "print(ptm_count_dict)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### STEP 7b create the PhosphoSite based Features files"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "feature_dir_psite = feature_dir + 'PHOSPHOSITE_PLUS/'\n",
    "ptm_feature_file_list, ptm_count_dict, ptm_feature_dict, mapping_dict = CoDIAC.PTM.write_PTM_features(Interpro_ID,\n",
    "                                                                                                      uniprot_reference_file,\n",
    "                                                                                                      feature_dir_psite,\n",
    "                                                                                                      mapping_file,\n",
    "                                                                                                      N_OFFSET,\n",
    "                                                                                                      C_OFFSET,\n",
    "                                                                                                      gap_threshold=0.7,\n",
    "                                                                                                      num_PTM_threshold=5,\n",
    "                                                                                                      PHOSPHOSITE_PLUS=True)\n",
    "print(\"Wrote these feature files:\")\n",
    "print(ptm_feature_file_list)\n",
    "print(\"These belong to the following fasta file:\")\n",
    "print(output_fasta)  #comes from block above - the short header format of the fasta header\n",
    "print(ptm_count_dict)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#### Modified PTM feature files to remove modifications for TNS1, which has issues that have yet to be resolved.\n",
    "This involved Phosphserine, Phosphothreonine, and Phosphotyrosine for ProteomeScout.\n",
    "This involved Phosphoserine, Phosphothreonine, Phosphotyrosine, and Ubiquitination for PhosphositePlus\n"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "ptm_feature_file_list_psite = [\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features/PHOSPHOSITE_PLUS/IPR000719_Phosphotyrosine.feature',\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features/PHOSPHOSITE_PLUS/IPR000719_Ubiquitination.feature',\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features/PHOSPHOSITE_PLUS/IPR000719_Acetylation.feature',\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features/PHOSPHOSITE_PLUS/IPR000719_Phosphothreonine.feature',\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features/PHOSPHOSITE_PLUS/IPR000719_Phosphoserine.feature']\n",
    "ptm_feature_file_list_pscout = [\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features//ProteomeScout/IPR000719_Phosphotyrosine.feature',\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features//ProteomeScout/IPR000719_N6-acetyllysine.feature',\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features//ProteomeScout/IPR000719_Phosphothreonine.feature',\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features//ProteomeScout/IPR000719_Phosphoserine.feature',\n",
    "    'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features//ProteomeScout/IPR000719_Ubiquitination.feature']\n",
    "ptm_feature_file_list_all = ptm_feature_file_list_psite + ptm_feature_file_list_pscout"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Create the annotation tracks for PTM features and \n",
    "for feature_file in ptm_feature_file_list_all:\n",
    "    file_arr = feature_file.split('.')\n",
    "    annotation_file = file_arr[0] + '.ann'\n",
    "    CoDIAC.jalviewFunctions.print_ann_file(feature_file, alignment_file, annotation_file)\n",
    "\n",
    "\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "#create an integration of all PTMs into one file.\n",
    "feature_file = feature_dir + 'PTM_all_features.feature'\n",
    "feature_combined, feature_color_dict = CoDIAC.jalviewFunctions.combine_feature_files(feature_file,\n",
    "                                                                                     ptm_feature_file_list)\n",
    "\n",
    "annotation_file = feature_dir + 'PTM_all_features_promals3d.ann'\n",
    "jalviewFunctions.print_ann_file(feature_file, alignment_file, annotation_file)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "### Step 8 combine feature files from ProteomeScout and PhosphoSitePlus and generate annotation tracks."
   ]
  },
  {
   "cell_type": "markdown",
   "metadata": {},
   "source": [
    "#paired list\n",
    "\n"
   ]
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "pairs = {}\n",
    "\n",
    "proteomescout_base = 'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features/ProteomeScout/IPR000719_'\n",
    "PSP_base = 'Data/Uniprot_Reference/Features_relative_to_reference/PTM_features/PHOSPHOSITE_PLUS/IPR000719_'\n",
    "pairs['N6-acetyllysine'] = [proteomescout_base + 'N6-acetyllysine.feature', PSP_base + 'Acetylation.feature']\n",
    "pairs['Phosphoserine'] = [proteomescout_base + 'Phosphoserine.feature', PSP_base + 'Phosphoserine.feature']\n",
    "pairs['Phosphothreonine'] = [proteomescout_base + 'Phosphothreonine.feature', PSP_base + 'Phosphothreonine.feature']\n",
    "pairs['Phosphotyrosine'] = [proteomescout_base + 'Phosphotyrosine.feature', PSP_base + 'Phosphotyrosine.feature']\n",
    "pairs['Ubiquitination'] = [proteomescout_base + 'Ubiquitination.feature', PSP_base + 'Ubiquitination.feature']\n",
    "\n",
    "output_dir = 'Data/Features/PTMS_all/'\n",
    "new_feature_files = {}\n",
    "\n",
    "for mod in pairs.keys():\n",
    "    feature_file = output_dir + 'Prot_kinase_dom_IPR000719_' + mod + '.feature'\n",
    "    feature_combined, feature_color_dict = CoDIAC.jalviewFunctions.combine_feature_files(feature_file, pairs[mod])\n",
    "    new_feature_files[mod] = feature_file\n"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# also combine the phosphoserine and phosphothreonine sets\n",
    "feature_file = output_dir + 'Prot_kinase_dom_IPR000719_Phosphoserine_Phosphothreonine.feature'\n",
    "feature_combined, feature_color_dict = CoDIAC.jalviewFunctions.combine_feature_files(feature_file,\n",
    "                                                                                     ['Data/Features/PTMS_all/Prot_kinase_dom_IPR000719_Phosphothreonine.feature',\n",
    "                                                                                      'Data/Features/PTMS_all/Prot_kinase_dom_IPR000719_Phosphoserine.feature'])"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "# Changed the Phosphoserine and Phosphothreonine feature names to Phosphorylation(ST)\n",
    "feature_file = output_dir + 'Prot_kinase_dom_IPR000719_Phosphoserine_Phosphothreonine.feature'\n",
    "jalviewFunctions.print_ann_file(feature_file, alignment_file,\n",
    "                                output_dir + 'Prot_kinase_dom_IPR000719_Phosphoserine_Phosphothreonine.ann')"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "#replaced the combined/merged features that overlapped in both resources and then run the annotation tracks\n",
    "for mod in new_feature_files.keys():\n",
    "    feature_file = new_feature_files[mod]\n",
    "    annotation_file = feature_file.replace('.feature', '.ann')\n",
    "    jalviewFunctions.print_ann_file(feature_file, alignment_file, annotation_file)"
   ],
   "outputs": [],
   "execution_count": null
  },
  {
   "cell_type": "code",
   "metadata": {},
   "source": [
    "#Combine all PTMs into one final file \n",
    "# updated the files for all PTMs to use just one modification name. These are in PTMs_final\n",
    "ptm_feature_all_dir = 'Data/Features/PTMs_all/PTMs_final/'\n",
    "\n",
    "#run the combined jalview annotation\n",
    "ST_file = ptm_feature_all_dir + 'Prot_kinase_dom_IPR000719_Phosphoserine_Phosphothreonine.feature'\n",
    "jalviewFunctions.print_ann_file(ST_file, alignment_file, ST_file.replace('.feature', '.ann'))\n",
    "\n",
    "feature_file_list = []\n",
    "for mod in ['N6-acetyllysine', 'Phosphotyrosine', 'Phosphoserine', 'Phosphothreonine', 'Ubiquitination']:\n",
    "    feature_file = ptm_feature_all_dir + 'Prot_kinase_dom_IPR000719_' + mod + '.feature'\n",
    "    jalviewFunctions.print_ann_file(feature_file, alignment_file, feature_file.replace('.feature', '.ann'))\n",
    "    feature_file_list.append(feature_file)\n",
    "output_final_PTMs_all = ptm_feature_all_dir + 'Prot_kinase_dom_IPR000719_PTMs_all.feature'\n",
    "feature_combined, feature_color_dict = CoDIAC.jalviewFunctions.combine_feature_files(output_final_PTMs_all,\n",
    "                                                                                     feature_file_list)\n",
    "jalviewFunctions.print_ann_file(output_final_PTMs_all, alignment_file,\n",
    "                                output_final_PTMs_all.replace('.feature', '.ann'))"
   ],
   "outputs": [],
   "execution_count": null
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.9.16"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 2
}
